index.js

```diff
@@ -119,16 +99,80 @@
   //  $$type :: String
   var $$type = '@@type';
 
-  //  type :: Any -> String
+  //  RPARSE :: RegExp
+  var RPARSE = /^(?:((?!\.|_)@?[0-9a-zA-Z./_%-]{1,214}):)?([^]+?)(?:@(\d+))?$/;
+
+  //       TypeIdentifier :: (Nullable String, String, Number)
+  //                      -> TypeIdentifier
+  function TypeIdentifier(namespace, name, version) {
+    return {
+      namespace: namespace,
+      name: name,
+      version: version
+    };
+  }
+
+  //       parseTypeIdentifier :: String -> TypeIdentifier
+  function parseTypeIdentifier(s) {
+    if (!s.length) return TypeIdentifier(null, s, 1);
+    var parsed = RPARSE.exec(s);
+    return TypeIdentifier(
+      parsed[1] || null,
+      parsed[2],
+      parsed[3] ? parseInt(parsed[3], 10) : 1
+    );
```

Comment:
I'd like to avoid type coercion here too:

```javascript
return TypeIdentifier(
  parsed[1] == null ? null : parsed[1],
  parsed[2],
  parsed[3] == null ? 1 : parseInt(parsed[3], 10)
);
```